mava/components/jax/training/trainer.py

```diff
@@ -301,6 +338,22 @@ def on_building_init_end(self, builder: SystemBuilder) -> None:
 
         builder.store.networks = builder.store.network_factory()
 
+        builder.store.policy_opt_states = {}
```

Comment:
Why is this code duplicated three times? See the other comments.

Reply:
In general, most of the code is repeated so I refactored Trainer.py to have most of the functionality in the Base Class